contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,348 @@
+other Kubernetes APIs as possible. So we define the selector for bulk operations
+as following:
+
+```
+// TODO: Find a better name for it?
+type BulkSelector struct {
+	SingleTypeBulkSelector []SingleTypeBulkSelector
+}
+
+// TODO: Find a better name for it?
+type SingleTypeBulkSelector struct {
+	// FIXME: This should be meta.k8s.io/GroupVersionKind instead (or something
+	// similar to it).
+	ResourceType string
+
+	// We would like to reuse the same ListOptions definition as we are using
+	// in regular APIs. For the purpose of bulk operations, you can just specify
+	// more than one.
+	//
+	// NOTE: In the initial implementation, we will require all Options (for a
+	// given resource type) to have exactly the same ResourceVersion - otherwise
+	// we will reject it at the validation level.
+	// This will simplify the initial implementation as we need to guarantee that
+	// watch events within a single resource types are delivered with increasing
+	// resource versions, which may be more tricky with different resource versions.
+	Options []ListOptions
+}
+```
+
+ With such API in place, all List/Watch requests coming from kubelets will be
+allowed to List/Watch only objects that are "attached" to pods bound to a given
+node. We will create a dedicated admission plugin that will be responsible for
+detecting whether a given kubelet (user in general) is allowed to list/watch
+objects requested by `BulkSelector ` (the exact mechanism for it is out of
+scope for this doc) and either rejecting the whole request or letting it go (we
+are not going to support partial rejections (e.g. for this resource you can
+proceed for that one no).
+
+ With the above semantics, whenever user wants to start watching another object
+(or set of objects) or drop one (or a set), he needs to break the watch and
+initiate a new one with different selectors. That isn't perfect, but we still
+can solve all usecases with it.
+ What is most important is that a bunch of them can still be solved efficiently.
+As an example, we can solve watching secrets in Kubelet as following:
+- create a store that will also support the following checks on add/update/delete
+operations: "if the resource version of object is older that we is currently
+cached, ignore it"
+- use this store underneath the reflector framework (though reflector will
+require some changes)
+- whenever a set of secrets to watch changes:
+  - break the current watch
+  - retrieve current state of secrets that are new to be watched and add it to
+    the local cache
+  - remove from local cache secrets that are no longer supposed to be watched
+  - send new watch request with the updated filter (with RV from the moment when
+    we broke the watch)
+
+With that approach, even if watch is lagging and we will get some previous
+version of secret (after retrieving it), it will be ignored.
+
+TODO: We should also allow for breaking the watch if the ACLs change (and kubelet
```

Comment:
I think this is required for an initial implementation.  Bounded time (as mentioned above) would be fine.